### ISSUE TITLE
ModuleList and get_subdict naming

### ISSUE THREAD

**Issue by raymond00000** [tristandeleu/pytorch-meta#114]
if I create a ModuleList  with this. so for each MetaLinearNorm, it is anonymous. 

            self.**layers** = nn.ModuleList(
            [MetaLinearNorm(in_size, out_size, bias=False)
             for (in_size, out_size) in zip(in_sizes, sizes)])

then in forward, suppose I should change to this?

          for linear in self.**layers**:
            x = F.dropout(F.relu(linear(x, params=self.get_subdict(params, '**layers**'))), p=0.5, training=True)

but module list is a list of MetaLinearNorm. will using the same **layers** fine? it seems not okay?

**Comment by tristandeleu** [comment]
There is no module called `MetaLinearNorm`, but I imagine you mean `MetaLinear`. You can use `nn.ModuleList` without any issue, you just need to get the index of the layer where you want the extract the `params` from. The `params` is just a dictionary containing the parameters, and the keys are the same keys as the ones you'd find in `model.named_parameters()` (or `model.meta_named_parameters()` if you have a mix of PyTorch modules and meta-modules). Here is an example:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

from torchmeta.modules import MetaModule, MetaLinear
from torchmeta.utils.gradient_based import gradient_update_parameters

class MLP(MetaModule):
    def __init__(self):
        super(MLP, self).__init__()
        hidden_sizes = [(3, 5), (5, 7)]
        self.layers = nn.ModuleList([MetaLinear(*io_sizes)
            for io_sizes in hidden_sizes])

    def forward(self, inputs, params=None):
        outputs = inputs
        for i, layer in enumerate(self.layers):
            outputs = F.relu(layer(outputs,
                params=self.get_subdict(params, f'layers.{i}')))
        return outputs

model = MLP()

inputs = torch.rand(2, 3)
targets = torch.randint(7, size=(2,))

inner_outputs = model(inputs)
inner_loss = F.cross_entropy(inner_outputs, targets)

params = gradient_update_parameters(model, inner_loss)

outer_outputs = model(inputs, params=params)
outer_loss = F.cross_entropy(outer_outputs, targets)
```

**Comment by raymond00000** [comment]
Many thanks for quick reply and example code. may I ask one more question? 

if part of my model contains some frozen parts, for example, a lstm

for example, 

        for param in model.lstm.parameters():
            param.requires_grad = False

 can I still call this way? or I have to replace the model to more specific one: model.layers.

      params = gradient_update_parameters(**model**, inner_loss)


thanks a lot for advice.

**Comment by tristandeleu** [comment]
No it should not require any further changes: the `gradient_update_parameters` function calls `torch.autograd.backward` under the hood, so any parameter which has been frozen will not be adapted. Same thing happens for any parameters in a standard PyTorch Module, they will not get updated with this function (see the ANIL example to learn more about it).

**Comment by raymond00000** [comment]
ok, i will try further according to the ANIL. 

may i ask one more question on dataset part. this one is a simple custom dataset.

    import torch
    import torch.utils.data as Data
    import numpy as np

    test = np.array([0,1,2,3,4,5,6,7,8,9,10,11])

    test = np.array([0,0,0,1,1,1,2,2,2,3,3,3])

    inputing = torch.tensor(np.array([test[i:i + 3] for i in range(10)]))
    target = torch.tensor(np.array([test[i:i + 1] for i in range(10)]))

    torch_dataset = Data.TensorDataset(inputing,target)
    batch = 3

    loader = Data.DataLoader(
        dataset=torch_dataset,
        batch_size=batch,
        collate_fn=lambda x:(
            torch.cat(
                [x[i][j].unsqueeze(0) for i in range(len(x))], 0
                ).unsqueeze(0) for j in range(len(x[0]))
            )
        )

    for (i,j) in loader:
        print(i)
        print(j)
        print("")

the result is this. 

     tensor([[[0, 0, 0],
         [0, 0, 1],
         [0, 1, 1]]])
     tensor([[[0],
         [0],
         [0]]])

     tensor([[[1, 1, 1],
         [1, 1, 2],
         [1, 2, 2]]])
     tensor([[[1],
         [1],
         [1]]])

tensor([[[2, 2, 2],
         [2, 2, 3],
         [2, 3, 3]]])
tensor([[[2],
         [2],
         [2]]])

tensor([[[3, 3, 3]]])
tensor([[[3]]])

I suppose I have to convert the dataset for BatchMetaDataLoader?

     dataset = omniglot("data", ways=5, shots=5, test_shots=15, meta_train=True, download=True)
    dataloader = BatchMetaDataLoader(dataset, batch_size=16, num_workers=4)


I tried to study the OmniglotDataset, but I could not figure out the way. 

    OmniglotDataset(Dataset) -> return (image, target)

   OmniglotClassDataset(ClassDataset) -> based on index, return OmniglotDataset
     
    Omniglot(CombinationMetaDataset)  -> init with dataset , num_classes_per_task
       -> sample for num_classese from self.dataset[class_ind] (ClassDataset)

Many thanks for advice!

**Comment by tristandeleu** [comment]
As I said in #113, it really depends on what you want to do (it depends on what your tasks are, if they are created by combining data from different labels like in a few-shot classification setting). If you do want something similar to your output, you could make it so that each instance of your `Task` class returns the 3x3 input tensor, and target tensor.
